# Job Data Exploration

## Objective
Understand the struture and quality of jobs_practice_2 before building a pipeline

## Source
- jobs_practice_2.csv

## Import relevant libraries and data

In [8]:
import pandas as pd 

job_data = pd.read_csv("../data/jobs_practice_2.csv")
job_data

,job_id,order_date,customer_name,region,product_line,units_completed,revenue,material_cost,labor_cost,freight_cost,rush_flag,discount_pct,status,quoted_ship_date,actual_ship_date,rework_hours
0,2001,2/12/2026,Valley Med,Northeast,Implants,57,8213.08,2485.21,939.94,822.80,N,5.0,Shipped,2/20/2026,2/23/2026,2.1
1,2002,2/25/2026,Acme Health,Midwest,Kits,342,41522.11,21070.40,5604.65,384.16,N,0.0,Cancelled,3/5/2026,NaN,0.0
2,2003,2/24/2026,Valley Med,Midwest,Components,306,43906.96,14198.38,5039.45,591.27,N,2.5,Cancelled,3/6/2026,NaN,0.0
3,2004,3/11/2026,Metro Ortho,West,Custom Assemblies,319,0.00,18257.42,3340.47,198.70,N,2.5,Cancelled,3/20/2026,NaN,0.0
4,2005,2/14/2026,Beacon Surgical,West,Implants,80,8631.69,2758.61,972.56,842.95,N,10.0,Shipped,2/22/2026,3/3/2026,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
75,2076,1/17/2026,Valley Med,South,Kits,41,6207.89,2993.42,742.12,616.05,N,5.0,In Production,1/27/2026,NaN,3.3
76,2077,3/1/2026,Summit Care,Northeast,Components,318,24537.92,9162.73,4011.28,475.56,N,2.5,Shipped,3/8/2026,3/14/2026,7.3
77,2078,3/1/2026,Valley Med,Northeast,Custom Assemblies,74,4133.13,1512.97,930.62,730.95,Y,7.5,Shipped,3/9/2026,3/7/2026,3.7
78,2079,2/11/2026,North Clinic,Midwest,Custom Assemblies,354,56375.05,17627.61,7585.25,896.69,Y,10.0,Shipped,2/24/2026,2/28/2026,2.7


## Schema Validation (Structure Check)

Questions
- Are columns named consistently?
- Are numbers actually numeric?
- Are dates strings?

In [11]:
job_data.head(5)

,job_id,order_date,customer_name,region,product_line,units_completed,revenue,material_cost,labor_cost,freight_cost,rush_flag,discount_pct,status,quoted_ship_date,actual_ship_date,rework_hours
0,2001,2/12/2026,Valley Med,Northeast,Implants,57,8213.08,2485.21,939.94,822.80,N,5.0,Shipped,2/20/2026,2/23/2026,2.1
1,2002,2/25/2026,Acme Health,Midwest,Kits,342,41522.11,21070.40,5604.65,384.16,N,0.0,Cancelled,3/5/2026,NaN,0.0
2,2003,2/24/2026,Valley Med,Midwest,Components,306,43906.96,14198.38,5039.45,591.27,N,2.5,Cancelled,3/6/2026,NaN,0.0
3,2004,3/11/2026,Metro Ortho,West,Custom Assemblies,319,0.00,18257.42,3340.47,198.70,N,2.5,Cancelled,3/20/2026,NaN,0.0
4,2005,2/14/2026,Beacon Surgical,West,Implants,80,8631.69,2758.61,972.56,842.95,N,10.0,Shipped,2/22/2026,3/3/2026,0.0


In [12]:
job_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   job_id            80 non-null     int64  
 1   order_date        80 non-null     object 
 2   customer_name     80 non-null     object 
 3   region            80 non-null     object 
 4   product_line      80 non-null     object 
 5   units_completed   80 non-null     int64  
 6   revenue           80 non-null     float64
 7   material_cost     80 non-null     float64
 8   labor_cost        80 non-null     float64
 9   freight_cost      80 non-null     float64
 10  rush_flag         80 non-null     object 
 11  discount_pct      80 non-null     float64
 12  status            80 non-null     object 
 13  quoted_ship_date  80 non-null     object 
 14  actual_ship_date  53 non-null     object 
 15  rework_hours      80 non-null     float64
dtypes: float64(6), int64(2), object(8)
memory usag

### Observations
- Columns are named consistently
- Numeric columns look like they are coming in as numerics
- Dates are coming in as text 

## Null / missing data check

Questions
- Which columns are critical?
- Which columns contain Nulls?
    - Are nulls acceptable here?

In [18]:
# Returns the number of null values for each column
job_data.isnull().sum()

job_id               0
order_date           0
customer_name        0
region               0
product_line         0
units_completed      0
revenue              0
material_cost        0
labor_cost           0
freight_cost         0
rush_flag            0
discount_pct         0
status               0
quoted_ship_date     0
actual_ship_date    27
rework_hours         0
dtype: int64

### Observations
- actual_ship_date has 27 null values
    - this might be ok because thouse jobs might not have shipped yet


## Uniqueness & keys

Questions
- Are there any primary keys?
    - If so, what are they?
- Are there any duplicate records?

In [19]:
job_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   job_id            80 non-null     int64  
 1   order_date        80 non-null     object 
 2   customer_name     80 non-null     object 
 3   region            80 non-null     object 
 4   product_line      80 non-null     object 
 5   units_completed   80 non-null     int64  
 6   revenue           80 non-null     float64
 7   material_cost     80 non-null     float64
 8   labor_cost        80 non-null     float64
 9   freight_cost      80 non-null     float64
 10  rush_flag         80 non-null     object 
 11  discount_pct      80 non-null     float64
 12  status            80 non-null     object 
 13  quoted_ship_date  80 non-null     object 
 14  actual_ship_date  53 non-null     object 
 15  rework_hours      80 non-null     float64
dtypes: float64(6), int64(2), object(8)
memory usag

In [20]:
# How many unique values exist
job_data["job_id"].nunique()

80

In [21]:
# How many duplicate values exists
job_data.duplicated(subset=["job_id"]).sum()

np.int64(0)

### Observations

- job_id might be a primary key
- There are 80 unique values in the job_id column
- There are 0 duplicated values in the job_id column
- Because there are 80 unique values in the job_id column and there are 80 rows in the data set, this would be a valid primary key
-   The .duplicated() check also confirms this, returning 0 for the number of duplicated values

## Distribution of numeric data

Questions
- Are there any negative values?
    - If so, should they exist?
- Are there any outliers?
- Are there any weird ranges?

In [23]:
job_data.describe()

,job_id,units_completed,revenue,material_cost,labor_cost,freight_cost,discount_pct,rework_hours
count,80.0000,80.000000,80.000000,80.000000,80.000000,80.000000,80.000000,80.000000
mean,2040.5000,239.625000,25075.763625,10901.498375,3523.080250,465.424625,3.375000,3.787500
std,23.2379,131.774392,18901.829174,7572.888027,2302.020202,263.794273,3.647871,2.949292
min,2001.0000,31.000000,0.000000,801.500000,257.000000,48.220000,0.000000,0.000000
25%,2020.7500,125.500000,8445.445000,3809.917500,1413.722500,232.935000,0.000000,1.125000
50%,2040.5000,259.500000,21875.960000,9503.665000,3326.570000,449.305000,2.500000,3.900000
75%,2060.2500,339.500000,40479.355000,15902.027500,4870.017500,700.887500,5.000000,5.725000
max,2080.0000,440.000000,69323.890000,30219.800000,9485.090000,899.890000,10.000000,11.000000


### Observations

- There is no missing numeric data
    - No null values in numeric columns
    - Dataset is complete for numeric fields
- Revenue includes some 0 values
    - min = 0
    - This could be ok, some jobs might not produce any revenue for a multitude of reasons
    - Checked: Reason is that they are cancelled orders
- Wide spread in revenue (possible skew)
    - mean is about 25,075
    - median (50%) is about 21,875
    - mean > median
        - right-skewed distribution
        - some high-value jobs are pulling the average up
    - Possible outliers
- Units completed has a large variation
    - min = 31
    - max = 440
    - mean is about 240
    - Big spread in production volume
    - Likely different job types or product lines
- Costs scale with revenue (Good Sign)
    - The largest costs also correspond with the largest revenues
        - No large costs with lower revenue
    - No obvious inconsistency
        - High revenue + 0 costs (would be bad data)
- discount_pct looks reasonable
    - min = 0
    - max = 10
    - mean = 3.37
    - Discounts look to be within expected range
        - No weird values (like 50% or 200%)
- Rework hours look low (good operational signal)
    - min = 0
    - max = 11
    - mean is about 3.78
    - Rework is relatively small compared to production
    - No extreme outliers
- Freight costs seem consistent
    - min is about 48
    - max is about 899
    - mean is about 465


#### Observation Checks

In [ ]:
# Revenue that = 0 
job_data[job_data["revenue"] == 0]

# reason is for cancelled orders

,job_id,order_date,customer_name,region,product_line,units_completed,revenue,material_cost,labor_cost,freight_cost,rush_flag,discount_pct,status,quoted_ship_date,actual_ship_date,rework_hours
3,2004,3/11/2026,Metro Ortho,West,Custom Assemblies,319,0.0,18257.42,3340.47,198.70,N,2.5,Cancelled,3/20/2026,NaN,0.0
12,2013,1/12/2026,Valley Med,Midwest,Components,399,0.0,15091.73,3553.45,64.75,N,7.5,Cancelled,1/23/2026,NaN,0.0
39,2040,2/3/2026,Summit Care,South,Implants,266,0.0,5049.03,4813.54,636.82,N,2.5,Cancelled,2/19/2026,NaN,0.0
62,2063,3/2/2026,Metro Ortho,West,Custom Assemblies,437,0.0,23230.20,4061.00,731.20,N,10.0,Cancelled,3/13/2026,NaN,0.0
69,2070,1/25/2026,Summit Care,Midwest,Kits,36,0.0,2089.64,257.00,131.69,N,5.0,Cancelled,2/22/2026,NaN,0.0


## Category Checks

Questions
- Are there any typos?
- Are there any unexpected categories

Categories:
- region
- product_line
- rush_flag
- status

In [ ]:
job_data["region"].unique() 

array(['Northeast', 'Midwest', 'West', 'South'], dtype=object)

In [34]:
job_data["product_line"].unique()

array(['Implants', 'Kits', 'Components', 'Custom Assemblies'],
      dtype=object)

In [35]:
job_data["rush_flag"].unique()

array(['N', 'Y'], dtype=object)

In [ ]:
job_data["status"].unique()

array(['shipped', 'cancelled', 'in production'], dtype=object)

### Observations

Categories look to be uniform 